src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs

```diff
@@ -164,60 +285,68 @@ public ForContentLength(bool keepAlive, long contentLength, Frame context)
+                var limit = (int)Math.Min(_inputLength, int.MaxValue);
                 if (limit == 0)
                 {
                     _context.RequestFinished();
-                    return new ValueTask<int>(0);
+                    return new ValueTask<ArraySegment<byte>>();
                 }
 
-                var task = _context.SocketInput.ReadAsync(buffer.Array, buffer.Offset, limit);
+                var task = _context.SocketInput.PeekAsync();
 
                 if (task.IsCompleted)
                 {
                     // .GetAwaiter().GetResult() done by ValueTask if needed
-                    var actual = task.Result;
-                    _inputLength -= actual;
+                    var actual = Math.Min(task.Result.Count, limit);
 
-                    if (actual == 0)
+                    if (task.Result.Count == 0)
                     {
                         _context.RejectRequest(RequestRejectionReason.UnexpectedEndOfRequestContent);
                     }
 
-                    if (_inputLength == 0)
+                    if (task.Result.Count < _inputLength)
                     {
-                        _context.RequestFinished();
+                        return task;
+                    }
+                    else
+                    {
+                        var result = task.Result;
+                        var part = new ArraySegment<byte>(result.Array, result.Offset, (int)_inputLength);
+                        return new ValueTask<ArraySegment<byte>>(part);
                     }
-
-                    return new ValueTask<int>(actual);
                 }
                 else
                 {
-                    return new ValueTask<int>(ReadAsyncAwaited(task.AsTask()));
+                    return new ValueTask<ArraySegment<byte>>(PeekAsyncAwaited(task));
                 }
             }
 
-            private async Task<int> ReadAsyncAwaited(Task<int> task)
+            private async Task<ArraySegment<byte>> PeekAsyncAwaited(ValueTask<ArraySegment<byte>> task)
             {
-                var actual = await task;
-                _inputLength -= actual;
+                var segment = await task;
 
-                if (actual == 0)
+                if (segment.Count == 0)
                 {
                     _context.RejectRequest(RequestRejectionReason.UnexpectedEndOfRequestContent);
                 }
 
-                if (_inputLength == 0)
+                if (segment.Count <= _inputLength)
                 {
-                    _context.RequestFinished();
+                    return segment;
+                }
+                else
+                {
+                    return new ArraySegment<byte>(segment.Array, segment.Offset, (int)_inputLength);
                 }
+            }
+
+            protected override void ConsumedBytes(int count)
+            {
+                _inputLength -= count;
 
-                return actual;
+                base.ConsumedBytes(count);
```

Comment:
I don't see the problem with putting implementation code in the base virtual method, but I'm not attached to it.

What do you suggest naming the abstract method called by ConsumedBytes()? OnConsumedBytes()?